File: python/linked_list/q234.py
"""
Given a singly linked list, determine if it is a palindrome.
"""
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def isPalindrome(self, head):
        """
        :type head: ListNode
        :rtype: bool
        """
        rev = None
        slow = fast = head
        while fast and fast.next:
            fast = fast.next.next
            rev, rev.next ,slow = slow, rev, slow.next
        if fast:
            slow = slow.next
        while slow != None:
            if slow.val != rev.val:
                return False
            slow = slow.next
            rev = rev.next
        return True

File: c++/linked_list/q234.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    bool isPalindrome(ListNode* head) {
        if(head == nullptr || head->next == nullptr){
            return true;
        }

        ListNode* fast = head; ListNode* slow = head;
        while(fast != nullptr && fast->next != nullptr){
            slow = slow->next;
            fast = fast->next->next;
        }


        if(fast)
            slow = slow->next;

        slow = reverseList(slow);

        while(slow != nullptr){
            if(head->val != slow->val)
                return false;
            head = head->next;
            slow = slow->next;
        }

        return true;
    }
private:
    ListNode* reverseList(ListNode* node){
        ListNode* prev = node;
        ListNode* curr = node->next;
        prev->next = nullptr;

        while(curr != nullptr){
            ListNode* temp = curr->next;
            curr->next = prev;
            prev = curr;
            curr = temp;
        }

        return prev;
    }
};
